udtube/data/conllu.py

```diff
@@ -3,25 +3,110 @@
 This is roughly compatible with the third-party package `conllu`, though it
 only has features we care about."""
 
+from __future__ import annotations
+
 import collections
+import dataclasses
 import re
 
-from typing import Dict, Iterable, Iterator, Optional, TextIO, Tuple
+from typing import Dict, Iterable, Iterator, List, Optional, TextIO, Tuple
+
+
+class Error(Exception):
+
+    pass
 
 
-# From: https://universaldependencies.org/format.html.
-_fieldnames = [
-    "id",
-    "form",
-    "lemma",
-    "upos",
-    "xpos",
-    "feats",
-    "head",
-    "deprel",
-    "deps",
-    "misc",
-]
+class ID:
+    """Representation of a CoNLL-U sentence ID.
+
+    Most of the time this is just a single integer, but MWEs use two integers
+    to represent a span of tokens.
+
+    Args:
+        lower (int): the lower or sole index.
+        upper (int, optional): the upper index, for MWEs.
+    """
+
+    lower: int
+    upper: int
+
+    def __init__(self, lower: int, upper: Optional[int] = None):
+        self.lower = lower
+        self.upper = self.lower if upper is None else upper
```

Comment:
It seems insanely unlikely to happen (impossible maybe?), but in the nuclear event that it does, I would like to handle the case if lower > upper. Maybe just in the spirit of fail-fast, a raise here.

Reply:
done